update_data.py

```diff
@@ -106,6 +107,26 @@ def split_pwgs(pwgs_string):
     """
     return [x.strip().upper() for x in pwgs_string.split(',') if x.strip()]
 
+def update_notes(sql_args, mcm_request, existing_sample, existing_request_prepid, request_changed):
```

Comment:
I guess you are doing this to have a couple of nice small functions instead of _that one_ that does **everything** which is very good! However, I would prefer "update_notes" function taking care only of notes and not being responsible for updating request in McM (because request update might include other updates, like interested PWGs) (also, if interested PWGs changed, but notes didn't, `update_notes` will not actually do anything with _notes_, but still update request in McM, because interested PWGs changed). One last small thing, which is not a bad code per se, but I don't like `existing_sample[5]` and `sql_args[19]` in this function. Now simple `update_notes` function is very aware of what is happening "outside", i.e. if we added another item to `sql_args` and notes became 21st element, `update_notes` would also have to be updated to [20] even though it has nothing to do with `sql_args` and should not even care about other items in `sql_args` list.

I would keep
```
if request_changed:
    logger.info('Updating %s', existing_request_prepid)
    response = mcm.update('requests', mcm_request)
    logger.info('Updated %s: %s', existing_request_prepid, response)
```
In the `insert_or_update` function (i.e. where it was before) and have two functions: `update_notes` and `update_interested_pwgs`. These two functions would return two values: boolean if value changed AND the new value if it changed or old value if note. `update_notes` would look like:
```
def update_notes(mcm_request, samples_notes):
    mcm_notes = mcm_request.get('notes')
    if mcm_notes != samples_notes:
        return True, samples_notes.strip()

    return False, mcm_notes
```
`update_interested_pwgs` would look like:
```
def update_interested_pwgs(mcm_request, current_interested_pwgs, original_interested_pwgs):
    # A bunch of set operations
    # Something something something
    if original_interested_pwgs_string != new_interested_pwgs_string:
        return True, new_interested_pwgs_string

    return False, original_interested_pwgs_string
```

And I would call these functions like:
```
notes_changed, sql_args[19] = update_notes(mcm_request, existing_sample[5])
interested_pwgs_changed, sql_args[16] = update_interested_pwgs(mcm_request, current_interested_pwgs, original_interested_pwgs)
sql_args[17] = sql_args[16]  # Update reference for the next sync with McM
if notes_changed or interested_pwgs_changed:
    # Update in McM...
```
This way `update_notes` and `update_interested_pwgs` do just one thing, do not care about `sql_args` or `existing_sample` and return whether we need to update or not. What do you think?

Reply:
Thanks for the detailed response, I'm really learning a lot. I did change the functions but don't know how to trigger an update to test if things are working fine? :grimacing: